### ISSUE TITLE
Silently skip over any unnecessary BOM at the start of input if the effective encoding is UTF-8

### ISSUE THREAD

**Issue by paul-breen** [paul-breen/xcsv#7]
Some text editors/tools (mainly on Windows) will add an unnecessary BOM to the start of UTF-8 encoded text.  The Unicode UTF-8 standard neither requires nor recommends a BOM for UTF-8 encoded text, and states that a BOM should be treated as a zero-width non-breaking space.

Pandas itself silently skips over any BOM at the start of input, otherwise it messes up its parsing.  This is exactly the same issue we have in parsing the extended header section.

Taking into account the UTF-8 standard and Pandas behaviour, and based on our work in [cassava](https://github.com/paul-breen/cassava/issues/22) around essentially the same issue, we should silently skip any BOM present at the start of the extended header section (which is normally the start of the XCSV input), if the effective encoding is UTF-8.

Python provides the (pseudo) encoding `utf-8-sig` (in addition to `utf-8`) for silently skipping a BOM.  If a user specifies this encoding, then the input will still be parsed correctly, it just won't be done by XCSV but by the Python `open()` call.   However, I feel that insisting users provide this encoding to manage this is not reasonable because:

* it's a bit subtle
* at the moment, there's no way to provide an alternative encoding via the CLI tools (the packages support it though)
* moreover though, doing this will mean that any file opened for writing (e.g. XCSV.Writer, XCSV.File.write()) will write a BOM to the start of the output, thereby propagating this problematic behaviour (if users require this for their workflow, then they can explicitly specify this encoding in their code, or save the output afterwards in a tool that will add a BOM)

In conclusion then, as a BOM is not part of the data and will mess up our parsing of the extended header section, we should silently skip over it.  Pandas will skip over any BOM when parsing the data table (in the case where an XCSV file has no extended header section, or where an XCSV file has been concatenated from a separate header file, and a CSV file, and both had separately had a BOM written to the start of their input).
